File: web/serve.py
#!/usr/bin/env python3
"""Static server for testing this folder locally, with the headers the kernel needs.

The kernel is multi-threaded WebAssembly on SharedArrayBuffer, which browsers only
enable in a cross-origin isolated page, so plain `python -m http.server` is not enough.

    python serve.py [port]      # default 8080, then open http://localhost:8080/

Use http://localhost or http://127.0.0.1: other hostnames over plain http are not a
secure context, so isolation stays off.
"""
import os
import sys
from functools import partial
from http.server import SimpleHTTPRequestHandler, ThreadingHTTPServer

ROOT = os.path.dirname(os.path.abspath(__file__))


class Handler(SimpleHTTPRequestHandler):
    # compileStreaming() rejects anything but application/wasm.
    extensions_map = {**SimpleHTTPRequestHandler.extensions_map, ".wasm": "application/wasm", ".js": "text/javascript"}

    def end_headers(self):
        self.send_header("Cross-Origin-Opener-Policy", "same-origin")
        self.send_header("Cross-Origin-Embedder-Policy", "require-corp")
        self.send_header("Cross-Origin-Resource-Policy", "cross-origin")
        self.send_header("Cache-Control", "no-cache")
        super().end_headers()


if __name__ == "__main__":
    port = int(sys.argv[1]) if len(sys.argv) > 1 else 8080
    server = ThreadingHTTPServer(("", port), partial(Handler, directory=ROOT))
    print(f"serving {ROOT} at http://localhost:{port}/")
    try:
        server.serve_forever()
    except KeyboardInterrupt:
        pass

File: web/serve.mjs
#!/usr/bin/env node
// Tiny static server for testing release/ locally with the headers the kernel
// needs (cross-origin isolation, application/wasm). No dependencies.
//
//   node serve.mjs [port]      # default 8080, then open http://localhost:8080/
import { createServer } from "node:http";
import { createReadStream } from "node:fs";
import { stat } from "node:fs/promises";
import { extname, join, normalize, resolve, sep } from "node:path";
import { fileURLToPath } from "node:url";

const root = resolve(fileURLToPath(new URL(".", import.meta.url)));
const port = Number(process.argv[2]) || 8080;

const types = {
  ".html": "text/html; charset=utf-8",
  ".js": "text/javascript; charset=utf-8",
  ".mjs": "text/javascript; charset=utf-8",
  ".css": "text/css; charset=utf-8",
  ".json": "application/json",
  ".wasm": "application/wasm",
};

createServer(async (req, res) => {
  const headers = {
    "Cross-Origin-Embedder-Policy": "require-corp",
    "Cross-Origin-Opener-Policy": "same-origin",
    "Cross-Origin-Resource-Policy": "cross-origin",
    "Cache-Control": "no-cache",
  };
  try {
    // Not `new URL(req.url)`: a request target like "//" parses as a host and throws.
    let path = normalize(decodeURIComponent(req.url.split(/[?#]/)[0]));
    if (path.endsWith(sep) || path.endsWith("/")) path = join(path, "index.html");
    const file = join(root, path);
    if (file !== root && !file.startsWith(root + sep)) throw Object.assign(new Error(), { code: "ENOENT" });
    const info = await stat(file);
    if (!info.isFile()) throw Object.assign(new Error(), { code: "ENOENT" });

    const range = /^bytes=(\d*)-(\d*)$/.exec(req.headers.range ?? "");
    const type = types[extname(file)] ?? "application/octet-stream";
    if (range && (range[1] || range[2])) {
      const start = range[1] ? Number(range[1]) : Math.max(info.size - Number(range[2]), 0);
      const end = range[1] && range[2] ? Math.min(Number(range[2]), info.size - 1) : info.size - 1;
      if (start > end || start >= info.size) {
        res.writeHead(416, { ...headers, "Content-Range": `bytes */${info.size}` }).end();
        return;
      }
      res.writeHead(206, {
        ...headers,
        "Content-Type": type,
        "Accept-Ranges": "bytes",
        "Content-Range": `bytes ${start}-${end}/${info.size}`,
        "Content-Length": end - start + 1,
      });
      createReadStream(file, { start, end }).pipe(res);
      return;
    }
    res.writeHead(200, {
      ...headers,
      "Content-Type": type,
      "Accept-Ranges": "bytes",
      "Content-Length": info.size,
    });
    createReadStream(file).pipe(res);
  } catch (error) {
    res.writeHead(error.code === "ENOENT" ? 404 : 500, headers).end(error.code === "ENOENT" ? "not found" : "error");
  }
}).listen(port, () => console.log(`serving ${root} at http://localhost:${port}/`));
